From: Nikola Pajkovsky <npajkovsky@suse.cz>
Date: Tue, 29 Aug 2017 13:59:04 +0200
Subject: [PATCH] scsi: aacraid: report -ENOMEM to upper layer from
 aac_convert_sgraw2()
References: bsc#1049519,FATE#322889
Git-commit: 96676246981b8321fb7bcaf51147f3be7c436af5
Patch-Mainline: v4.14-rc1

aac_convert_sgraw2() kmalloc memory and return -1 on error, which should
be -ENOMEM. However, nobody is checking return value, so with this
change, -ENOMEM is propagated to upper layer.

Signed-off-by: Nikola Pajkovsky <npajkovsky@suse.cz>
Reviewed-by: Raghava Aditya Renukunta <RaghavaAditya.Renukunta@microsemi.com>
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 drivers/scsi/aacraid/aachba.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/scsi/aacraid/aachba.c b/drivers/scsi/aacraid/aachba.c
index d55de4b..a64285a 100644
--- a/drivers/scsi/aacraid/aachba.c
+++ b/drivers/scsi/aacraid/aachba.c
@@ -3944,8 +3944,12 @@ static long aac_build_sgraw2(struct scsi_cmnd *scsicmd,
 			if (!err_found)
 				break;
 		}
-		if (i > 0 && nseg_new <= sg_max)
-			aac_convert_sgraw2(rio2, i, nseg, nseg_new);
+		if (i > 0 && nseg_new <= sg_max) {
+			int ret = aac_convert_sgraw2(rio2, i, nseg, nseg_new);
+
+			if (ret < 0)
+				return ret;
+		}
 	} else
 		rio2->flags |= cpu_to_le16(RIO2_SGL_CONFORMANT);
 
@@ -3969,7 +3973,7 @@ static int aac_convert_sgraw2(struct aac_raw_io2 *rio2, int pages, int nseg, int
 
 	sge = kmalloc(nseg_new * sizeof(struct sge_ieee1212), GFP_ATOMIC);
 	if (sge == NULL)
-		return -1;
+		return -ENOMEM;
 
 	for (i = 1, pos = 1; i < nseg-1; ++i) {
 		for (j = 0; j < rio2->sge[i].length / (pages * PAGE_SIZE); ++j) {
-- 
1.8.5.6

